feat(button-toggle): create button-toggle based on MDC Web

annieyw opened the pull request.

mmalerba commented on src/material-experimental/mdc-button-toggle/button-toggle.ts:
```diff
@@ -0,0 +1,672 @@
+
+  constructor(private _changeDetector: ChangeDetectorRef,
+              private _elementRef: ElementRef<HTMLElement>) {
+    this._foundation = new MDCSegmentedButtonFoundation(this._adapter);
```
I notice we're not using many of the foundation methods: `selectSegment`, `unselectSegment`, `handleSelected`, `isSegmentSelected`, `getSelectedSegments`. Should we be? If not, what exactly are we getting out of using the foundation?

annieyw replied:
I'm a little confused on how to incorporate the foundation methods as the inputs and outputs are of `SegmentDetail` and I think the original API deals directly with `MatButtonToggle` components etc

mmalerba replied:
It looks like `SegmentDetail` is just the index an selected state. You could turn a list of `ButtonToggle` into a list of `SegmentDetail` like this:

```
toggles.toArray().map((t, i) => ({index: i, selected: t.selected}))
```

And you can convert a `SegmentDetail` to a `ButtonToggle` like this:

```
toggles.toArray()[detail.index]
```

annieyw commented on src/material-experimental/mdc-button-toggle/button-toggle.spec.ts:
```diff
@@ -0,0 +1,1042 @@
+import {dispatchMouseEvent} from '@angular/cdk/testing/private';
```
Not really, mostly just changing selectors because the structure is a little different now.
Added tests for the new role attributes

annieyw commented:
- Removed foundation because there wasn't much added utility
- Added test harness
  - Had to copy over tests instead of use the `*.shared.spec` because the selector is now different `<mat-button-toggle>` -> `<button mat-button-toggle>`

annieyw commented on src/material-experimental/mdc-button-toggle/_button-toggle-theme.scss:
```diff
@@ -0,0 +1,5 @@
+@use '@material/ripple/common';
+@use '@material/segmented-button/styles';
```
Yea the dev-app has the appropriate styles. I think the `@use` rule loads the functions/mixins automatically

mmalerba replied:
It does, and it seems like its also emitting styles, which we don't want. Importing this file should not cause styles to be emitted. It should just define mixins that can be used by the app to include the appropriate styles

annieyw commented on src/material-experimental/mdc-button-toggle/button-toggle.scss:
```diff
@@ -0,0 +1,5 @@
+@use '@material/ripple/common';
+@use '@material/segmented-button/styles';
```
I believe it has both because the theme is included here https://github.com/material-components/material-components-web/blob/master/packages/mdc-segmented-button/segment/_segment.scss#L33

mmalerba replied:
We don't want both here, this should be only the structural styles (see other MDC-based components for reference)

annieyw commented:
- Added appropriate styles from mdc mixins

Follow up:
- ripple
- universal app
- e2e test
- docs

mmalerba commented on src/material-experimental/mdc-button-toggle/_button-toggle-theme.scss:
```diff
@@ -0,0 +1,54 @@
+@mixin mat-mdc-button-toggle-typography($config-or-theme) {
+  $config: mat-get-typography-config($config-or-theme);
+  @include mat-using-mdc-typography($config) {
+    @include segment.core-styles($query: $mat-typography-styles-query);
```
same todo here as color

mmalerba commented on src/material-experimental/mdc-button-toggle/button-toggle.scss:
```diff
@@ -0,0 +1,10 @@
+@import '../mdc-helpers/mdc-helpers';
+@import '../../cdk/a11y/a11y';
+
+@include segment.core-styles($query: $mat-base-styles-query);
```
same todo as theme file for these

mmalerba commented on src/material-experimental/mdc-button-toggle/testing/button-toggle-group-harness.ts:
```diff
@@ -0,0 +1,48 @@
+
+  /** Gets whether the button toggle group is laid out vertically. */
+  async isVertical(): Promise<boolean> {
+    return (await this.host()).hasClass('mat-button-toggle-vertical');
```
probably just just be `return false`?

mmalerba commented on src/material-experimental/mdc-button-toggle/testing/button-toggle-harness.ts:
```diff
@@ -0,0 +1,116 @@
+    return (await this.host()).getAttribute('aria-labelledby');
+  }
+
+  // /** Gets the appearance that the button toggle is using. */
```
you can either: 
- uncomment this, mark it deprecated, and `return 'standard'`
- just delete it

mmalerba commented on src/material-experimental/mdc-theming/_all-theme.scss:
```diff
@@ -27,6 +28,7 @@
     @include mat-mdc-core-theme($theme-or-color-config);
     @include mat-mdc-autocomplete-theme($theme-or-color-config);
     @include mat-mdc-button-theme($theme-or-color-config);
+    @include mat-mdc-button-toggle-theme($theme-or-color-config);
```
did you verify that this is applying the typography styles as well?

annieyw replied:
Yes when I removed the typography styles from the theme file the text looked different

mmalerba commented:
@annieyw feel free to add the new package to the commit log checker in this PR

andrewseguin commented:
This PR fell out of date and the future of button toggle in MDC is unclear. We can use this as a reference if it gets ready for us to use again
